### ISSUE TITLE
Expanding/unflattening objects and access control

### ISSUE THREAD

**Issue by cwebber** [swicg/general#9]
This is a confused deputy attack which applies to any ActivityPub server that hosts multiple users and does unflattening for client convenience, but it isn't necessarily unique to ActivityPub; any social network system that does unflattening, has access control features, and has multiple users on the same system could hit this problem.  To a lesser extent, the attack may also be used when malicious users are not on the same server, but with some guidelines this is easier to prevent than when users are on the same server.

Say Alyssa and Ben are on different servers, and Alyssa has posted a private message to which Ben has access, and Mallet (who is on the same server as Ben but does not have access) but would like to  see it anyway.

Alyssa's post looks like:

```
{"@context": "https://www.w3.org/ns/activitystreams",
 "id": "https://social.example/alyssa/posts/fd055d64-fc13-4ae2-9c64-7a487df43b3b",
 "attributedTo": "https://social.example/alyssa/",
 "type": "Note",
 "to": ["https://social.example/alyssa/groups/friends/"],
 "content": "Looking forward to going to my family's haunted mansion this weekend!"}
```

Ben is a member of Alyssa's "friends" group, the message is delivered to his inbox.  When Ben's client retrieves his inbox, it conveniently unflattens it inside the collection.  This is done to reduce the number of round trips between Ben's client and the server; rather than making Ben's client fetch each individual item, all the information is already there.

```
{"@context": "https://www.w3.org/ns/activitystreams",
 "id": "https://chatty.example/ben/inbox",
 "type": "OrderedCollection",
 "first": {
    "type": "OrderedCollectionPage",
    "orderedItems": [
      {"id": "https://social.example/alyssa/posts/d404c9ca-bf39-4e8a-8731-1dcf738b74bd",
       "type": "Create",
       "author": {
         "type": "Person",
         "id": "https://social.example/alyssa/",
         "name": "Alyssa P. Hacker",
         "preferredUsername": "alyssa",
         "summary": "Lisp enthusiast hailing from MIT",
         "inbox": "https://social.example/alyssa/inbox/",
         "outbox": "https://social.example/alyssa/outbox/",
         "followers": "https://social.example/alyssa/followers/",
         "following": "https://social.example/alyssa/following/",
         "likes": "https://social.example/alyssa/likes/"},
       "object": {
         "type": "Note",
         "id": "https://social.example/alyssa/posts/fd055d64-fc13-4ae2-9c64-7a487df43b3b",
         "attributedTo": {
           "type": "Person",
           "id": "https://social.example/alyssa/",
           "name": "Alyssa P. Hacker",
           "preferredUsername": "alyssa",
           "summary": "Lisp enthusiast hailing from MIT",
           "inbox": "https://social.example/alyssa/inbox/",
           "outbox": "https://social.example/alyssa/outbox/",
           "followers": "https://social.example/alyssa/followers/",
           "following": "https://social.example/alyssa/following/",
           "likes": "https://social.example/alyssa/likes/"},
         "to": ["https://social.example/alyssa/groups/friends/"],
         "content": "Looking forward to going to my family's haunted mansion
                     this weekend!  The address is at 1900 Transylvania Avenue,
                     Spooksville, PA, USA."}},
      ...]}}
```

(Note that Ben's server has even recursively expanded Alyssa's profile as a courtesy!)

Somehow, Mallet is made aware of the presence of the object's id, but not its contents.  Perhaps someone has publicly liked the note, or it appears in some collection, or (less likely in this scenario with a uuid, but would have been easier if Alyssa was using a server like Mallet and Ben's with incrementing integer ids) perhaps Mallet can simply guess the uri of an upcoming post.  Regardless, Mallet does not have access to the object but would like to. Mallet realizes they are on the same server as someone who does have access, so they decide to trick their own server (which has cached the contents of Alyssa's message) into exposing the message's contents.

First, they post an activity which simply references the object and post it to their outbox:

```
{"@context": "https://www.w3.org/ns/activitystreams",
 "type": "Like",
 "object": "https://social.example/alyssa/posts/fd055d64-fc13-4ae2-9c64-7a487df43b3b"}
```

Then Mallet retrieves their outbox:

```
{"@context": "https://www.w3.org/ns/activitystreams",
 "id": "https://chatty.example/mallet/outbox",
 "type": "OrderedCollection",
 "first": {
    "type": "OrderedCollectionPage",
    "orderedItems": [
      {"type": "Like",
       "id": "https://chatty.example/mallet/p/85301",
       "author": {
         "type": "Person",
         "id": "https://chatty.example/mallet/",
         "name": "Mallet",
         "preferredUsername": "mallet",
         "summary": "Not a nice person.",
         "inbox": "https://chatty.example/mallet/inbox",
         "outbox": "https://chatty.example/mallet/outbox",
         "followers": "https://chatty.example/mallet/followers",
         "following": "https://chatty.example/mallet/following",
         "likes": "https://chatty.example/mallet/likes"}
       "object": {
         "type": "Note",
         "id": "https://social.example/alyssa/posts/fd055d64-fc13-4ae2-9c64-7a487df43b3b",
         "attributedTo": {
           "type": "Person",
           "id": "https://social.example/alyssa/",
           "name": "Alyssa P. Hacker",
           "preferredUsername": "alyssa",
           "summary": "Lisp enthusiast hailing from MIT",
           "inbox": "https://social.example/alyssa/inbox/",
           "outbox": "https://social.example/alyssa/outbox/",
           "followers": "https://social.example/alyssa/followers/",
           "following": "https://social.example/alyssa/following/",
           "likes": "https://social.example/alyssa/likes/"},
         "to": ["https://social.example/alyssa/groups/friends/"],
         "content": "Looking forward to going to my family's haunted mansion
                     this weekend!  The address is at 1900 Transylvania Avenue,
                     Spooksville, PA, USA."}}
    ...]}}
```

Now Mallet knows exactly where the haunted mansion is.  Mallet thinks they'll take a visit themselves.

Mallet and Ben's server did not have the full access control information of the remote object.  It had the object cached since it was able to retrieve it on Ben's behalf.  When a user requests an object in the client to server protocol, the server tries to reduce the number of requests by expanding contents.  Unfortunately, Mallet and Ben's server did not know the contents of Alyssa's friends list; it's not public information, even though it was used in delivery, since Alyssa would prefer to keep private whom she considers friends even to her own friends... it's her business, not theirs.  At any rate, the server was not trying hard to check anyway; it saw that it had a cached copy of the object, so it expanded it as a courtesy.

Unfortunately, this leaked private information that it should not have.

Access control is hard.  Combining access control with expansion is extra hard when you have multiple users, especially when access control is of remote objects, and your local server may not even be
privy as to who has access.

This scenario was described for ActivityPub, but I think any multi-user server doing expansion like this will face the same challenges.

I did not go into the details of remote users taking advantage of unflattening/expansion, which is also a risk.  There is a simple solution: never expand/unflatten non-public contents sent in a federated setting or otherwise retrieved outside the client to server protocol.

Following this, single-user servers, or servers that store the objects a user have access to independently (similar to how email servers have separate stores for each user generally) can have an easier time of it.  If you only have one user, and you only expand contents for that individual user, you won't have a problem.

Are there other ways to mitigate this?

**Comment by nightpool** [comment]
> and your local server may not even be privy as to who has access

This seems to be the core issue here. Why does the local server not know who has access? Shouldn't it create some sort of access control list for the object based on who received it in their outbox?

**Comment by jaywink** [comment]
If I understood this correctly, servers showing any content that is not public to someone who isn't clearly in the receiver end sounds like a serious (server) bug. This sounds more like implementation detail than actual protocol design issue? Server authors should ensure that private information is not leaked out in the same way they protect user accounts and passwords.

As an example, in Diaspora any non-public item can be viewed only if the user has "visibility" to it. This local visibility link is created when the item is created on the local server. This means access control is easy to implement to see if the user has access to the item. This is expensive but necessary. The downside is that visibility *cannot* be changed after the item has been created. Well, technically you could send out signed "add user X to item Y" but there is nothing like that in the protocol.

> Are there other ways to mitigate this?

IMHO servers should always ensure they only show content which they can reliably say the user should see. This might require expensive storage of visibility links or expensive on-demand lookups, but just "blurping" content to anyone with the URL is IMHO an implementation bug, not something that can be mitigated by protocol design. Of course there could be a mention as a security consideration that local servers MUST show non-public items only to users they can say received it?

**Comment by cwebber** [comment]
> If I understood this correctly, servers showing any content that is not public to someone who isn't clearly in the receiver end sounds like a serious (server) bug. This sounds more like implementation detail than actual protocol design issue? Server authors should ensure that private information is not leaked out in the same way they protect user accounts and passwords.
> [...]
> IMHO servers should always ensure they only show content which they can reliably say the user should see. This might require expensive storage of visibility links or expensive on-demand lookups, but just "blurping" content to anyone with the URL is IMHO an implementation bug, not something that can be mitigated by protocol design. Of course there could be a mention as a security consideration that local servers MUST show non-public items only to users they can say received it?

I agree, it would be a bug... and I think this isn't a protocol issue, as you said, it's a possible bug that a naive server could encounter.  I just realized that it would be an easy bug for me to myself implement incorrectly when I started looking at caching of expanded objects, and thought we should think about how to guide other servers to not also have the same problems.

**Comment by cwebber** [comment]
> This seems to be the core issue here. Why does the local server not know who has access? Shouldn't it create some sort of access control list for the object based on who received it in their inbox?

Yes, a reasonable approach to check whether a user can view something is whether the user has it in their inbox, and if not, see if you can retrieve it using their credentials.  If not to either, then they don't have access.
